File: triton-bot/src/main/python/publisher_consumer/module.py
import pickle
import exchange


class Module:
    CONNECTION_FACTORY_HOST = 'localhost'
    EXCHANGE_TYPE = 'fanout'

    def __init__(self):
        self.setup_channel()
        self.load_config()
        self.prepare()
        self.declare_exchanges()

    def setup_channel(self):
        connection = pika.BlockingConnection(
            pika.ConnectionParameters(self.CONNECTION_FACTORY_HOST))
        self.channel = connection.channel()

    def load_config(self):
        return

    def prepare(self):
        return

    def declare_exchanges(self, exchange):
        return

    def declare_publish(self, exchange):
        self.channel.exchange_declare(
            exchange=exchange.name, exchange_type=self.EXCHANGE_TYPE)

    def declare_consume(self, exchange, consumer):
        self.channel.exchange_declare(
            exchange=exchange.name, exchange_type=self.EXCHANGE_TYPE)

        queue_name = self.channel.queue_declare(
            queue='', exclusive=True).method.queue
        self.channel.queue_bind(exchange=exchange.name, queue=queue_name)

        self.channel.basic_consume(
            queue=queue_name, on_message_callback=consumer, auto_ack=True)
        self.channel.start_consuming()

    def publish(self, exchange, object):
        self.channel.basic_publish(
            exchange=exchange.name, routing_key='', body=pickle.dumps(object))

File: triton-soccer-ai/src/main/java/com/triton/modules/Module.java
package com.triton.modules;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.DeliverCallback;
import com.triton.messaging.Exchange;

import java.io.IOException;
import java.util.concurrent.TimeoutException;

import static com.triton.messaging.SimpleSerialize.simpleSerialize;

public abstract class Module extends Thread {
    private static final String CONNECTION_FACTORY_HOST = "localhost";
    private static final String EXCHANGE_TYPE = "fanout";

    private Channel channel;

    public Module() throws IOException, TimeoutException {
        setupChannel();
        loadConfig();
        prepare();
        declareExchanges();
    }

    private void setupChannel() throws IOException, TimeoutException {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(CONNECTION_FACTORY_HOST);
        Connection connection = factory.newConnection();
        channel = connection.createChannel();
    }

    protected void loadConfig() throws IOException {
    }

    protected void prepare() {
    }

    /**
     * Override to declare exchanges.
     *
     * @throws IOException
     * @throws TimeoutException
     */
    protected void declareExchanges() throws IOException, TimeoutException {
    }

    /**
     * Declares an exchange to publish to
     *
     * @param exchange the exchange
     * @throws IOException
     */
    protected void declarePublish(Exchange exchange) throws IOException {
        channel.exchangeDeclare(exchange.name(), EXCHANGE_TYPE);
    }

    /**
     * Declares an exchange to consume from. The messageConsumer function will be called when an message is received
     * from the exchange.
     *
     * @param exchange the exchange to consume from
     * @param callback the function to call once a message is received
     * @throws IOException
     */
    protected void declareConsume(Exchange exchange, DeliverCallback callback) throws IOException {
        channel.exchangeDeclare(exchange.name(), EXCHANGE_TYPE);
        String queueName = channel.queueDeclare().getQueue();
        channel.queueBind(queueName, exchange.name(), "");
        channel.basicConsume(queueName, true, callback, consumerTag -> {
        });
    }

    /**
     * Publishes to an exchange
     *
     * @param exchange the exchange to publish to
     * @param object   the object to send
     * @throws IOException
     */
    protected void publish(Exchange exchange, Object object) throws IOException {
        channel.basicPublish(exchange.name(), "", null, simpleSerialize(object));
    }
}
